File: day12/day12.cpp
#include <fstream>
#include <iostream>
#include <vector>

#include <fmt/format.h>

using namespace std;

static vector<vector<int>> load_graph(istream& input)
{
	vector<vector<int>> graph;
	string line;
	while (getline(input, line))
	{
		enum {NODE, DEPS} state = NODE;
		size_t pos = 0;
		size_t next;
		vector<int> deps;
		do
		{
			next = line.find(" ", pos);
			if (state == NODE)
			{
				if (line.rfind("<->", pos) == pos)
				{
					state = DEPS;
				}
			}
			else if (state == DEPS)
			{
				deps.push_back(stoi(line.substr(pos, next-pos)));
			}
			pos = next + 1;
		} while (next != string::npos);
		graph.emplace_back(move(deps));
	}
	return graph;
}

static void dfs(const vector<vector<int>>& g, int i, vector<bool>& visited)
{
	visited[i] = true;
	for (auto j: g[i])
	{
		if (!visited[j])
		{
			dfs(g, j, visited);
		}
	}
}

static size_t part1(const vector<vector<int>>& g)
{
	vector<bool> visited(g.size(), false);
	dfs(g, 0, visited);
	return count(visited.begin(), visited.end(), true);
}

static size_t part2(const vector<vector<int>>& g)
{
	vector<bool> visited(g.size(), false);
	size_t groups = 0;
	for (size_t i = 0; i < g.size(); i++)
	{
		if (!visited[i])
		{
			dfs(g, i, visited);
			groups++;
		}
	}
	return groups;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	auto graph = load_graph(input);
	input.close();

	cout << fmt::format("Part1: {}", part1(graph)) << endl
	     << fmt::format("Part2: {}", part2(graph)) << endl;
	return 0;
}

File: day12/day12.py
#!/usr/bin/env python3

def parse(txt):
    g = []
    for row in txt.splitlines():
        _, deps = row.split(" <-> ")
        g.append([int(i) for i in deps.split(", ")])
    return g

def dfs(g, i, visited):
    visited[i] = True
    for j in g[i]:
        if not visited[j]:
            dfs(g, j, visited)

def part1(g):
    visited = [False] * len(g)
    dfs(g, 0, visited)
    return sum(1 for i in visited if i)

def part2(g):
    visited = [False] * len(g)
    group = 0
    for i in range(len(g)):
        if not visited[i]:
            dfs(g, i, visited)
            group += 1

    return group

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    g = parse(txt)
    print("Part1:", part1(g))
    print("Part2:", part2(g))
